jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,188 @@
+
+import logging
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+__all__ = ["BadpixSelfcalStep"]
+
+
+class BadpixSelfcalStep(Step):
+    """
+    BadpixSelfcalStep: Flags residual artifacts as bad pixels in the DQ array
+    of a JWST exposure using a median filter and percentile cutoffs.
+
+    All input exposures in the association file (or manually-provided bkg_list) are combined
+    into a single background model using a MIN operation. The bad pixels are then identified
+    using a median filter and percentile cutoffs, and applied to the science data by setting
+    the flagged pixels, errors, and variances to NaN
+    and the DQ flag to DO_NOT_USE + OTHER_BAD_PIXEL.
+    """
+
+    class_alias = "badpix_selfcal"
+    bkg_suffix = "badpix_selfcal"
+
+    spec = """
+    flagfrac_lower = float(default=0.001)  #fraction of pixels to flag on the low-flux end
+    flagfrac_upper = float(default=0.001)  #fraction of pixels to flag on the high-flux end
+    kernel_size = integer(default=15)  #size of kernel for median filter
+    force_single = boolean(default=False)  #force single input exposure
+    skip = boolean(default=True)
+    """
+
+    def process(self, input, selfcal_list=[]):
+        """
+        Flag residual artifacts as bad pixels in the DQ array of a JWST exposure
+
+        Parameters
+        ----------
+        input: JWST data model or association
+            input science data to be corrected, or tuple of (sci, bkg, selfcal)
+
+        selfcal_list: list of user-defined ImageModels or filenames to use for selfcal
+
+        Returns
+        -------
+        output: JWST data model or association
+            data model with CRs flagged
+
+        Notes
+        -----
+        If selfcal_list is specified manually, it overrides any non-science exposures
+        in the association file.
+        If selfcal_list is set to None and an association file is read in, all exposures in the
+        association file, including science, background, and selfcal exposures,
+        are included in the MIN frame from which outliers are detected.
+        If selfcal_list is set to None and input is a single science exposure, the step will
+        be skipped with a warning unless the force_single parameter is set True.
+        In that case, the input exposure will be used as the sole background exposure,
+        i.e., true self-calibration.
+        """
+
+        input_sci, selfcal_list, bkg_list_asn = _parse_inputs(input, selfcal_list)
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No selfcal or background exposures provided for self-calibration. \
+                        Skipping step. If you want to force self-calibration with the science \
+                        exposure alone (generally not recommended), set force_single=True.")
```

Comment:
These warnings are not printing the way I think you intended - they show up with lots of spaces where the \ appears.  This should work better:
```suggestion
            log.warning("No selfcal or background exposures provided for self-calibration. "
                        "Skipping step. If you want to force self-calibration with the science "
                        "exposure alone (generally not recommended), set force_single=True.")
```

The rest of the warnings in this module should be similarly updated.